examples/xgboost/custom/xgboost_learner.py

```diff
@@ -0,0 +1,211 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from abc import abstractmethod
+
+import xgboost as xgb
+from sklearn.metrics import roc_auc_score
+from torch.utils.tensorboard import SummaryWriter
+
+from nvflare.apis.dxo import DXO, DataKind, from_shareable
+from nvflare.apis.fl_constant import FLContextKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.learner_spec import Learner
+from nvflare.app_common.app_constant import AppConstants
+
+
+class XGBoostLearner(Learner):
+    def __init__(
+        self,
+        train_task_name: str = AppConstants.TASK_TRAIN,
+    ):
+        super().__init__()
+        # Currently we support boosting 1 tree per round
+        # could further extend
+        self.trees_per_round = 1
+        self.train_task_name = train_task_name
+
+    def initialize(self, parts: dict, fl_ctx: FLContext):
+        # when a run starts, this is where the actual settings get initialized for learner
+
+        # set the paths according to fl_ctx
+        engine = fl_ctx.get_engine()
+        ws = engine.get_workspace()
+        app_dir = ws.get_app_dir(fl_ctx.get_job_id())
+
+        # get and print the args
+        fl_args = fl_ctx.get_prop(FLContextKey.ARGS)
+        self.client_id = fl_ctx.get_identity_name()
+        self.log_info(
+            fl_ctx,
+            f"Client {self.client_id} initialized with args: \n {fl_args}",
+        )
+
+        # set local tensorboard writer for local training info of current model
+        self.writer = SummaryWriter(app_dir)
+        # set the training-related contexts, this is task-specific
+        self.train_config(fl_ctx)
+
+    def train_config(self, fl_ctx: FLContext):
+        """Traning configurations customized to individual tasks
+        This can be specified / loaded in any ways
+        as long as they are made available for further training and validation
+        some potential items include but not limited to:
+
+        self.data_path
+        self.local_model_path
+        self.global_model_path
+        self.site_id_index_mapping
+
+        self.lr
+        self.objective
+        self.max_depth
+        self.eval_metric
+        self.nthread
+
+        self.train_X
+        self.train_y
+        self.valid_X
+        self.valid_y
+        """
+        raise NotImplementedError
```

Comment:
For all the instance variables "self.XXX" that you need in the `train` method.
We should require who ever implements this base class to set those up.

I would suggest in this PR you rename this file to `xgboost_higgs_learner.py` and copy the "train_config" codes from your `xgboost_higgs_learner.py`  to here.

We can think of how to abstract this out later when we want to move this into app_common.

Reply:
sure